src/silly_sprites/opengl/shader_program.cpp

```diff
@@ -0,0 +1,179 @@
+#include "shader_program.hpp"
+#include "error.hpp"
+#include "gsl/gsl"
+#include "utils.hpp"
+
+namespace sly::gl {
+    static constexpr auto fallback_vertex_source = R"(
+            #version 330 core
+            layout (location = 0) in vec3 aPos;
+
+            void main()
+            {
+                gl_Position = vec4(aPos.x, aPos.y, aPos.z, 1.0);
+            }
+        )";
+    static constexpr auto fallback_fragment_source = R"(
+            #version 330 core
+            out vec4 FragColor;
+            
+            void main()
+            {
+                FragColor = vec4(255.0f, 0.0f, 255.0f, 1.0f);
+            }    
+        )";
+
+    [[nodiscard]] constexpr std::string_view ShaderProgram::get_name_from_type(Type const type) {
+        switch (type) {
+            case Type::Vertex:
+                return "VERTEX";
+            case Type::Geometry:
+                return "GEOMETRY";
+            case Type::Fragment:
+                return "FRAGMENT";
+        }
+        return "INVALID";
+    }
+
+    [[nodiscard]] GLuint ShaderProgram::compile(Type const type, std::string_view const source) {
+        // compile
+
+        auto error_message = [&](GLuint id) -> std::string {
+            auto message = std::string{};
+            auto len = GLint{};
+            glGetShaderiv(id, GL_INFO_LOG_LENGTH, &len);
+            message.resize(gsl::narrow_cast<usize>(len - 2));
+            glGetShaderInfoLog(id, len - 1, nullptr, message.data());
+            return message;
+        };
+        auto compile_single = [&](GLuint& id, char const* single_source) -> GLint {
+            id = glCreateShader(sly::to_underlying(type));
+            glShaderSource(id, 1, &single_source, nullptr);
+            glCompileShader(id);
+            auto success = GLint{};
+            glGetShaderiv(id, GL_COMPILE_STATUS, &success);
+            return success;
+        };
```

Comment:
Style guide says to avoid output parameters.
The lambda should not return a resource that has to be manually cleaned up later (by calling `glDeleteShader()`). I'd suggest to either rewrite this somehow or to introduce an RAII wrapper type for a single shader that cleans up automatically.
```suggestion
        auto const compile_single = [&](std::string_view const single_source) -> tl::optional<GLuint> {
            auto const id = glCreateShader(sly::to_underlying(type));
            glShaderSource(id, 1, single_source.data(), nullptr);
            glCompileShader(id);
            auto result = GLint{};
            glGetShaderiv(id, GL_COMPILE_STATUS, &result);
            if (result == GL_TRUE) {
                return id;
            }
            return tl::nullopt;
        };
```